### PR TITLE
Gpl973 create or modify existing plate lookup endpoint

### PR DESCRIPTION
**Pull request by emrojo**
Closes GPL973

Changes proposed in this pull request:

* Add "pickable_samples" when argument "&include_samples=true" is added to the URL when using plate lookup by barcode

Coupled with https://github.com/sanger/lighthouse-ui/pull/80

### CODE REVIEW

File: lighthouse/blueprints/plates.py [hunk 1]
```diff
@@ -116,7 +116,9 @@ def find_plate_from_barcode() -> FlaskResponse:
         barcodes = request.args.getlist("barcodes[]")
         logger.debug(f"Barcodes to look for: {barcodes}")
 
-        plates = [format_plate(barcode) for barcode in barcodes]
+        include_samples = (request.args.get("include_samples") == "true")
```
Comment (pjvv) [inline lighthouse/blueprints/plates.py]: I think we should make this more generic, something similar to [this](https://developer.wordpress.org/rest-api/using-the-rest-api/global-parameters/#_fields) perhaps?

File: lighthouse/helpers/plates.py [hunk 2]
```diff
@@ -723,3 +727,17 @@ def format_plate(barcode: str) -> Dict[str, Union[str, bool, Optional[int]]]:
...
         "count_filtered_positive": count_filtered_positive if count_filtered_positive is not None else 0,
     }
+
+    if include_samples is True:
+        formated_response["pickable_samples"] = list(map(pickable_sample_attributes, fit_to_pick_samples))
+
+    return formated_response
+
+
+def pickable_sample_attributes(sample: Any) -> Any:
```
Comment (pjvv) [inline lighthouse/helpers/plates.py]: If possible, try to define the types more explicitly; using `Any` doesn't really give us any advantage when using type checking.

**Review by pjvv — CHANGES_REQUESTED** [review]

**Comment by pjvv** [comment]
You can either rebase or merge develop in since there are new changes there.

File: lighthouse/blueprints/plates.py (hunk 1 above)
Comment (emrojo) [reply lighthouse/blueprints/plates.py]: ok, seems a good change;  what do you propose?

File: lighthouse/helpers/plates.py [hunk 3]
```diff
@@ -693,17 +698,17 @@ def __sample_subject_for_dart_control_row(dart_control_row: Dict[str, str]) -> D
     }
 
 
-def format_plate(barcode: str) -> Dict[str, Union[str, bool, Optional[int]]]:
-    """Used by flask route /plates to format each plate. Determines whether there is sample data for the barcode and if
-    so, how many samples meet the fit to pick rules.
+def render_fields_format_plate(barcode: str) -> Dict[str, Callable[[], Union[str, bool, SampleDocs, Optional[int]]]]:
```
Comment (pjvv) [inline lighthouse/helpers/plates.py:701]: I find this name confusing/hard to read, what about `plate_field_wrapper`, `plate_fields`, `fields_for_plate`?

File: lighthouse/helpers/plates.py [hunk 4]
```diff
@@ -714,12 +719,56 @@ def format_plate(barcode: str) -> Dict[str, Union[str, bool, Optional[int]]]:
...
+    renderable = render_fields_format_plate(barcode)
+    formated_response = {}
+    for field in renderable:
+        # Not generate the field if is in the exclusion list
+        if not (field in exclude_props):
+            formated_response[field] = renderable[field]()
+    return formated_response
+
+
+def pickable_sample_attributes(sample: SampleDoc) -> SampleDoc:
```
Comment (pjvv) [inline lighthouse/helpers/plates.py:768]: Maybe add some comments here while the context is fresh regarding Biosero etc.?

**Review by pjvv — CHANGES_REQUESTED** [review]

File: lighthouse/helpers/plates.py (hunk 3 above)
Comment (emrojo) [reply lighthouse/helpers/plates.py:701]: Ok, going with field_generators_for_plate_lookup as I am thinking they are the fields+values

File: tests/helpers/test_plates_helpers.py [hunk 5]
```diff
@@ -1182,6 +1184,20 @@ def test_construct_cherrypicking_plate_failed_message_source_plates_not_in_mongo
...
 
+def test_format_plate(app, plates_lookup_without_samples, plates_lookup_with_samples):
+    with app.app_context():
+        assert (
+            format_plate("plate_123", exclude_props=["pickable_samples"]) == plates_lookup_without_samples["plate_123"]
+        )
+        assert format_plate("plate_123") == plates_lookup_with_samples["plate_123"]
+
+
+def test_render_fields_format_plate(app, plates_lookup_without_samples, plates_lookup_with_samples):
```
[the remark this answers is not in the document]
Comment (emrojo) [reply tests/helpers/test_plates_helpers.py:1195]: Test not right, correcting

File: tests/blueprints/test_plates.py [hunk 6]
```diff
@@ -104,17 +102,65 @@ def test_get_plates_endpoint_successful(app, client, samples, priority_samples,
     }
 
 
+def test_get_plates_endpoint_no_barcode_in_request(client):
+    response = client.get(
+        "/plates",
+        content_type="application/json",
```
Comment (pjvv) [inline tests/blueprints/test_plates.py:108]: Do we need to specify the content type?

**Review by pjvv — DISMISSED** [review]

File: lighthouse/blueprints/plates.py [hunk 7]
```diff
@@ -97,26 +98,38 @@ def create_plate_from_barcode() -> FlaskResponse:
...
     For example:
     To fetch data for the plates with barcodes '123', '456' and '789':
 
-    `GET /plates?barcodes[]=123&barcodes[]=456&barcodes[]=789`
+    `GET /plates?barcodes=123,456,789`
 
     This endpoint responds with JSON and the body is in the format:
 
-    `{"plates":[{"barcode":"123","plate_map":true,"number_of_fit_to_pick":0}]}`
+    `{"plates":[
```
[the remark this answers is not in the document]
Comment (emrojo) [reply lighthouse/blueprints/plates.py]: Updated coment

File: tests/blueprints/test_plates.py (hunk 6 above)
Comment (emrojo) [reply tests/blueprints/test_plates.py]: Yes not needed. Removed

**Pull request merged**